app/src/main/java/com/github/se/eduverse/ui/search/UserProfileScreen.kt

```diff
@@ -125,13 +129,41 @@ fun UserProfileScreen(
                       }
 
                   // Follow/Unfollow Button
-                  val currentUserId = FirebaseAuth.getInstance().currentUser?.uid
                   if (currentUserId != null && currentUserId != userId) {
                     Button(
-                        onClick = { TODO() },
-                        modifier = Modifier.fillMaxWidth().padding(horizontal = 16.dp)) {
-                          Text(if (profile.followers > 0) "Unfollow" else "Follow")
+                        onClick = { viewModel.toggleFollow(currentUserId, userId) },
+                        modifier =
+                            Modifier.fillMaxWidth()
+                                .padding(horizontal = 16.dp)
+                                .testTag("follow_button"),
+                        enabled = followActionState != FollowActionState.Loading) {
+                          when (followActionState) {
+                            FollowActionState.Loading -> {
+                              CircularProgressIndicator(
+                                  modifier = Modifier.size(24.dp),
+                                  color = MaterialTheme.colorScheme.onPrimary)
+                            }
+                            else -> {
+                              when (uiState) {
+                                is ProfileUiState.Success -> {
+                                  val profile = (uiState as ProfileUiState.Success).profile
+                                  Text(
+                                      if (profile.isFollowedByCurrentUser) "Unfollow" else "Follow")
```

Comment:
Here I think it's better to have the button text as a `MutableStateFlow` so that when uiState is `ProfileUiState.Error`, because of loss of connection for example the button can have the text corresponding to the latest state instead of always being Follow on error: then the code would update the mutable state if the uiState is `ProfileUiState.Success` and do nothing when it's `ProfileUiState.Error` and the button text would take the value of the mutable state flow